### PR TITLE
[python] fix trees_to_dataframe and enhance test

### PR DESCRIPTION
**Pull request by StrikerRUS**
- transfer `test_trees_to_dataframe` test from `test_basic` to `test_engine` with `skipIf` guard as we use `lgb.train()` there;
- add test for the edge case described in https://github.com/microsoft/LightGBM/pull/2592#discussion_r364435337;
- it turned out that dictionary keys are `list` in Python 2, so it's needed to convert them into a `set`

### CODE REVIEW

File: tests/python_package_test/test_engine.py [hunk 1]
```diff
@@ -1810,3 +1810,54 @@ def test_binning_same_sign(self):
         predicted = est.predict(new_x)
         self.assertNotAlmostEqual(predicted[0], predicted[1])
         self.assertAlmostEqual(predicted[1], predicted[2])
+
+    @unittest.skipIf(not lgb.compat.PANDAS_INSTALLED, 'pandas is not installed')
```
Comment (jameslamb) [inline tests/python_package_test/test_engine.py]: oh this is great!
